File: FunctorsAlem/sorting.hs
import Data.List

data Person = Person {name :: String, age :: Int} deriving Show
data Employee = Employee {tag :: Int, person :: Person} deriving Show

cmpAge :: Int -> Int -> Ordering
cmpAge x y | x < y  = LT
           | x > y  = GT
           | x == y = EQ
           
cmpPerson :: Person -> Person -> Ordering           
cmpPerson x y = cmpAge (age x) (age y)

cmpEmployee :: Employee -> Employee -> Ordering
cmpEmployee x y = cmpAge ((age. person) x) ((age.person) y)

cmap :: (b -> a) -> (a -> a -> c) -> (b -> b-> c)
cmap f c = \x y -> c (f x) (f y)

cmpPerson' = cmap age cmpAge
cmpEmployee' = cmap (age.person) cmpAge

p1 = Person "Olaf" 23
p2 = Person "George" 27
p3 = Person "Carl" 19

e1 = Employee 1234 p1
e2 = Employee 1235 p2
e3 = Employee 1236 p3

main = do
    print (sortBy cmpPerson [p1,p2,p3])
    print (sortBy cmpEmployee [e1,e2,e3])
    print (sortBy cmpPerson' [p1,p2,p3])
    print (sortBy cmpEmployee' [e1,e2,e3])

File: FunctorsAlem/sorting.cpp
#include <iostream>     // std::cout
#include <algorithm>    // std::sort
#include <vector>       // std::vector
#include <string>
#include <functional>

typedef struct person {
  int age;
  std::string name;
} person;

typedef struct employee {
  int id;
  person p;
} employee;

bool sortAge (int i,int j) { return (i<j); }

int getAge(person p) { return p.age; }
person getPerson(employee e) { return e.p; }

auto const compose = [](auto f, auto g) {
        return [f, g](auto x) {
                return g(f(x));
        };
};

template<class A, class B>
auto contramap(std::function<bool(B, B)> f, std::function<B(A)> g)
{
  return [f, g](A x, A y){return f(g(x), g(y));};
}

auto sortEmployee = contramap<employee, int>(sortAge, compose(getPerson, getAge));

int main () {
  person p1 = {23, "Olaf"};
  person p2 = {27, "George"};
  person p3 = {19, "Carl"};

  employee es[] = {{1, p1}, {2, p2}, {3, p3}};

  std::vector<employee> myvector (es, es+3);

  // using function as comp
  std::sort (myvector.begin(), myvector.end(), sortEmployee);

  // print out content:
  std::cout << "myvector contains:";
  for (std::vector<employee>::iterator it=myvector.begin(); it!=myvector.end(); ++it)
    std::cout << ' ' << (it->p).name << ' ' << (it->p).age;
  std::cout << '\n';

  return 0;
}
